tokio/src/runtime/shell.rs

```diff
@@ -55,29 +49,14 @@ impl Shell {
     }
 }
 
-fn clone_waker(ptr: *const ()) -> RawWaker {
-    let w1 = unsafe { ManuallyDrop::new(Arc::from_raw(ptr as *const Handle)) };
-    let _w2 = ManuallyDrop::new(w1.clone());
-
-    RawWaker::new(
-        ptr,
-        &RawWakerVTable::new(clone_waker, wake, wake_by_ref, drop_waker),
-    )
-}
-
-fn wake(ptr: *const ()) {
-    use crate::park::Unpark;
-    let unpark = unsafe { Arc::from_raw(ptr as *const Handle) };
-    (unpark).unpark()
-}
-
-fn wake_by_ref(ptr: *const ()) {
-    use crate::park::Unpark;
-
-    let unpark = ptr as *const Handle;
-    unsafe { (*unpark).unpark() }
-}
+impl Wake for Handle {
+    /// Wake by value
+    fn wake(self: Arc<Self>) {
+        Wake::wake_by_ref(&self);
+    }
 
-fn drop_waker(ptr: *const ()) {
-    let _ = unsafe { Arc::from_raw(ptr as *const Handle) };
+    /// Wake by reference
+    fn wake_by_ref(arc_self: &Arc<Self>) {
+        arc_self.0.unpark();
```

Comment:
nit: pretty sure `&Arc<Self>` is 9a valid self type](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=83332cf3f4f27aeca647499309d5a40c):
```suggestion
    fn wake_by_ref(self: &Arc<Self>) {
        self.0.unpark();
```

Reply:
If it is it must be recent. It was not a self type when we shipped so it would probably break MSRV. I can check tho

Reply:
I confirmed. On 1.39:

```
`&std::sync::Arc<runtime::shell::Handle>` cannot be used as the type of `self` without the `arbitrary_self_types` feature
```

Reply:
ah, nevermind!